lib/Runtime/Library/JavascriptRegularExpression.cpp

```diff
@@ -642,155 +663,94 @@ namespace Js
             RegexHelper::IsResultNotUsed(callInfo.Flags));
     }
 
-    bool JavascriptRegExp::HasObservableConstructor(RecyclableObject* instance, ScriptContext* scriptContext)
+    bool JavascriptRegExp::HasOriginalRegExType(RecyclableObject* instance)
     {
-        return HasObservableValue(
-            instance,
-            PropertyIds::constructor,
-            scriptContext->GetLibrary()->GetRegExpConstructor(),
-            scriptContext);
-    }
+        JavascriptLibrary* library = instance->GetLibrary();
 
-    bool JavascriptRegExp::HasObservableExec(RecyclableObject* instance, ScriptContext* scriptContext)
-    {
-        return HasObservableValue(
-            instance,
-            PropertyIds::exec,
-            scriptContext->GetLibrary()->GetRegexExecFunction(),
-            scriptContext);
+        if (instance->GetType() != library->GetRegexType())
```

Comment:
`JavascriptLibrary` instances are per `ScriptContext` so in the case that the instance is a marshalled object from a different `ScriptContext` we must get the right context's type.  I think marshalled objects return their owning context's types and such, but you should verify with a sanity check unit test.

Same for the built-in function objects that are used for testing below.